### PR TITLE
[2.3.0][Tune][Doc] Restructure API reference

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

Cherry pick of https://github.com/ray-project/ray/pull/32311 and https://github.com/ray-project/ray/pull/32442 (fixed link check errors from the 1st PR).

This PR splits up long API refs into individual pages, one dedicated to each method/class. There are now 3 layers: an index page (ex: `doc/tune/api/api.rst`), a summary page for each group of APIs (ex: `doc/tune/api/trainable.rst`), and an individual page for each API (ex: `doc/tune/api/docs/tune.Trainable.rst`).

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by cadedaniel** [comment]
Thanks @justinvyu ! Let's wait until https://github.com/ray-project/ray/pull/32446 is merged.

**Review by cadedaniel — APPROVED** [review]

**Comment by cadedaniel** [comment]
Ready for approval @zhe-thoughts. Not blocking the release test run today.

File: .gitignore [hunk 1]
```diff
@@ -119,6 +119,7 @@ scripts/nodes.txt
 /doc/_build
 /doc/source/_static/thumbs
 /doc/source/tune/generated_guides/
+/doc/source/**/doc/
```
Comment (cadedaniel) [inline .gitignore:122]: @justinvyu can you explain why this is necessary?

**Review by cadedaniel — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (justinvyu) [reply .gitignore:122]: We now have autogenerated API files under `/doc/source/tune/api/doc/*`, `/doc/source/data/api/doc/*`, etc. We shouldn't track these with git.

**Review by cadedaniel — APPROVED** [review]

**Comment by justinvyu** [comment]
Note: I had to make another commit removing a reference to an API that has landed in master but not 2.3.0.

**Comment by cadedaniel** [comment]
cc @zhe-thoughts merging this doc change in now

**Pull request merged**